config_parser.cc

```diff
@@ -199,12 +203,13 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
           new_config);
       config_stack.push(new_config);
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
+      end_block_counter++;
       if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
         // Error.
         break;
       }
       config_stack.pop();
-    } else if (token_type == TOKEN_TYPE_EOF) {
+    } else if (token_type == TOKEN_TYPE_EOF && start_block_counter == end_block_counter) {
```

Comment:
You might want to move this check into a separate `if`. As it is, if this condition is not met, the error would be because of an "invalid token" instead of an unmatched bracket.